### PR TITLE
Enable composite swap for `testpoolswap`

### LINKED ISSUE #938: Add `estimatecompositepaths` RPC
Current state: 

- Currently, `compositeswap` messages in the protocol require pool paths IDs. 
- In the node, the `compositeswap` RPC includes a small path finder implementation to resolve this first before broadcasting the tx.
- However, other implementations now have reimplement this. While this is intended and each implementation is free to implement a more efficient path finder algorithm, it's useful to expose the result of this so that it's easier for other implementations to reuse this if needed. 

To-Do: 

- Implement an `estimatecompositepaths` RPC that calculates provides list the path IDs as result for a given swap.

### LINKED ISSUE #939: Add `testcompositeswap` RPC
Consider implementing `testcompositeswap` for `compositeswap`, analogous to `testpoolswap` for `poolswap`.

### PR DESCRIPTION
**Pull request by mambisi**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:
Add `estimatecompositepaths` rpc endpoint for testing composite swaps
Add `testcompositeswap` rpc endpoint for testing composite swaps

#### Which issue(s) does this PR fixes?:

Fixes #938
Fixes #939

### CODE REVIEW

File: src/masternodes/rpc_poolpair.cpp [hunk 1]
```diff
@@ -1026,6 +1038,195 @@ UniValue testpoolswap(const JSONRPCRequest& request) {
...
+
+    }
+
+    return UniValue(res.msg);
+}
+
+UniValue estimatecompositepaths(const JSONRPCRequest& request) {
+    auto pwallet = GetWallet(request);
+
+    RPCHelpMan{"testcompositeswap",
```
Comment (prasannavl) [inline src/masternodes/rpc_poolpair.cpp:1145]: Help requires update.

File: src/masternodes/mn_checks.cpp [hunk 2]
```diff
@@ -3529,7 +3568,7 @@ Res CPoolSwap::ExecuteSwap(CCustomCSView& view, std::vector<DCT_ID> poolIDs) {
             }
             intermediateView.Flush();
 
-            return res;
+           return Res::Ok(swapAmountResult.ToString());;
```
Comment (Bushstar) [inline src/masternodes/mn_checks.cpp:3571]: Double semicolon

File: src/masternodes/rpc_poolpair.cpp [hunk 3]
```diff
@@ -1026,6 +1038,195 @@ UniValue testpoolswap(const JSONRPCRequest& request) {
...
+                                 {
+                                         {"txid", RPCArg::Type::STR_HEX, RPCArg::Optional::NO, "The transaction id"},
+                                         {"vout", RPCArg::Type::NUM, RPCArg::Optional::NO, "The output number"},
+                                 },
+                                },
+                        },
+                       },
+               },
+               RPCResult{
+                       "\"hash\"                  (string) The hex-encoded hash of broadcasted transaction\n"
```
Comment (Bushstar) [inline src/masternodes/rpc_poolpair.cpp:1077]: Incorrect result. Please update the return help test for compositeswap and estimatecompositepaths as well.

File: src/masternodes/rpc_poolpair.cpp [hunk 4]
```diff
@@ -1026,6 +1038,195 @@ UniValue testpoolswap(const JSONRPCRequest& request) {
...
+    {
+        LOCK(cs_main);
+        CCustomCSView dummy(*pcustomcsview); // create dummy cache for test state writing
+        // If no direct swap found search for composite swap
+        if (!dummy.GetPoolPair(poolSwapMsg.idTokenFrom, poolSwapMsg.idTokenTo)) {
+
+            auto compositeSwap = CPoolSwap(poolSwapMsg, targetHeight);
+            res = compositeSwap.TextExecuteSwap(dummy);
+
+            if (!res.ok) {
```
Comment (Bushstar) [inline src/masternodes/rpc_poolpair.cpp:1124]: You can use res as bool.

`if (!res)`

File: src/masternodes/rpc_poolpair.cpp [hunk 5]
```diff
@@ -1026,6 +1038,195 @@ UniValue testpoolswap(const JSONRPCRequest& request) {
...
+                    }
+                    errorMsg += ')';
+                }
+                throw JSONRPCError(RPC_INVALID_REQUEST, errorMsg);
+            }
+        }
+
+    }
+
+    return UniValue(res.msg);
```
Comment (Bushstar) [inline src/masternodes/rpc_poolpair.cpp:1139]: No need to wrap in UniValue, just return res.msg.

File: src/masternodes/rpc_poolpair.cpp (hunk 1 above)
Comment (mambisi) [reply src/masternodes/rpc_poolpair.cpp:1145]: Okay

File: src/masternodes/mn_checks.cpp (hunk 2 above)
Comment (mambisi) [reply src/masternodes/mn_checks.cpp:3571]: :+1:

File: src/masternodes/rpc_poolpair.cpp (hunk 3 above)
Comment (mambisi) [reply src/masternodes/rpc_poolpair.cpp:1077]: :+1:

File: src/masternodes/rpc_poolpair.cpp (hunk 4 above)
Comment (mambisi) [reply src/masternodes/rpc_poolpair.cpp:1124]: Okay

File: src/masternodes/rpc_poolpair.cpp (hunk 5 above)
Comment (mambisi) [reply src/masternodes/rpc_poolpair.cpp:1139]: Okay

File: src/masternodes/mn_checks.cpp [hunk 6]
```diff
@@ -3348,6 +3348,70 @@ bool IsMempooledCustomTxCreate(const CTxMemPool & pool, const uint256 & txid)
 
 std::vector<DCT_ID> CPoolSwap::CalculateSwaps(CCustomCSView& view) {
```
Comment (Bushstar) [inline src/masternodes/mn_checks.cpp:3349]: Please combine CalculateSwaps and TextExecuteSwap and toggle with a bool.

Comment (mambisi) [reply src/masternodes/mn_checks.cpp:3349]: Okay

File: src/masternodes/rpc_poolpair.cpp [hunk 7]
```diff
@@ -1010,19 +1042,51 @@ UniValue testpoolswap(const JSONRPCRequest& request) {
...
+                    errorMsg += ')';
+                }
+                throw JSONRPCError(RPC_INVALID_REQUEST, errorMsg);
+            }
+
+            for (const auto& id : poolIds) {
+                pools.push_back(id.ToString());
             }
+        } else {
+            CPoolPair pp = poolPair->second;
```
Comment (prasannavl) [inline src/masternodes/rpc_poolpair.cpp:1066]: Null reference when no pool pair.

**Pull request merged**